### PR TITLE
[cross_file] Unify XFile interface on web and mobile

### PR DESCRIPTION
**Pull request by ahmednfwela**
In cases where you try to extend `XFile` on web, you get a mismatch implementation error, and that's because of the `path` variable.
on web, it's defined with a getter AND a setter
```dart
late String path;
```
but on io (which is the same as `XFileBase`), it's defined with only a getter 
```dart
  @override
  String get path {
    return _file.path;
  }
```

so trying to extend `XFile` is not possible since on web it will try to extend the setter (which has no implmentation), and the best solution is to expose `XFileBase`.

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated `CHANGELOG.md` to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

**Comment by stuartmorgan** [comment]
> on web, it's defined with a getter AND a setter
> 
> ```dart
> late String path;
> ```
> 
> but on io (which is the same as `XFileBase`), it's defined with only a getter

This seems like the real bug; we should fix that rather than expose internals to patch around it. We should do a breaking change to remove the setter for the web unless there's some use case for it there that I'm missing.

/cc @ditman

**Comment by ahmednfwela** [comment]
@stuartmorgan I don't think making a breaking change is necessary here, we can just expose the interface and everything is solved

**Comment by stuartmorgan** [comment]
> we can just expose the interface and everything is solved

No, everything would not be solved:
1. The interfaces of the two implementations are supposed to be identical. Your PR doesn't fix that
2. There is reason for a path setter to be there (and the fact that the variable isn't also `final` looks like an accident). Its existence in the API surface is a mistake. Your PR doesn't fix that either.

> or we can also add an empty setter on io

That makes (2) worse instead of fixing it.

**Comment by ahmednfwela** [comment]
@stuartmorgan Is that good enough ?

File: packages/cross_file/lib/cross_file.dart [hunk 1]
```diff
@@ -2,4 +2,5 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+export 'src/types/base.dart';
```
[the remark this answers is not in the document]
Comment (ditman) [reply packages/cross_file/lib/cross_file.dart:5]: @ahmednfwela I thought this was the change you did to get XFile to work as you needed, and make it extensible? (by fixing the difference in API):

https://github.com/flutter/packages/pull/438/commits/4014258530e468aae78ad0a07d0924729b683887

(What I mean is: if by making the API consistent across mobile/web you can now extend XFile and build your memoizer, there's some things that need to change in this PR, like the title, not exporting XFileBase, the CHANGELOG entry so it reflects what the actual change was, etc...)

**Comment by ahmednfwela** [comment]
@ditman I think the PR is ready for merge now

**Comment by ahmednfwela** [comment]
yes, I have tested it.

**Comment by ditman** [comment]
Hey @ahmednfwela, I can see that this PR has been created in the `master` branch of your repo. That goes against step 5 of the [Tree Hygiene Overview](https://github.com/flutter/flutter/wiki/Tree-hygiene#overview).

Can you move this change to its own branch? That way I can clone and push freely to it without having to wrestle with multiple copies of `master`.

_(For example, I was going to try to fix the `format+analyze` check for you, but I won't do it directly in your `master` branch!)_

Thanks!

**Comment by ahmednfwela** [comment]
oh my bad, I am replacing this PR with a new one

**Pull request closed without merging**